From ab9ca938ef6cddf371838c745a3b2eb2de0d0625 Mon Sep 17 00:00:00 2001
From: Artur Wieczorek <artwik@wp.pl>
Date: Wed, 22 Apr 2015 20:19:06 +0200
Subject: Set proper file name characters encoding in the test unit.

File name characters encoding should be set properly for the file systems which use internally UTF-8 character encoding (like wxGTK) because otherwise Unicode file names are not converted properly (with wxString::fn_str) to the internal file name representation.

diff --git a/tests/test.cpp b/tests/test.cpp
index 97d12d7..610f4fd 100644
--- a/tests/test.cpp
+++ b/tests/test.cpp
@@ -405,6 +405,59 @@ static int TestCrtReportHook(int reportType, char *message, int *)
 
 #endif // wxUSE_VC_CRTDBG
 
+#ifdef __WXGTK__
+// The code to set encoding is taken from wxApp::Initialize.
+#include "wx/fontmap.h"
+
+void SetEncoding()
+{
+    // gtk+ 2.0 supports Unicode through UTF-8 strings
+    wxConvCurrent = &wxConvUTF8;
+#ifdef __UNIX__
+    // decide which conversion to use for the file names
+
+    // (1) this variable exists for the sole purpose of specifying the encoding
+    //     of the filenames for GTK+ programs, so use it if it is set
+    wxString encName(wxGetenv(wxT("G_FILENAME_ENCODING")));
+    encName = encName.BeforeFirst(wxT(','));
+    if (encName.CmpNoCase(wxT("@locale")) == 0)
+        encName.clear();
+    encName.MakeUpper();
+    if (encName.empty())
+    {
+#if wxUSE_INTL
+        // (2) if a non default locale is set, assume that the user wants his
+        //     filenames in this locale too
+        encName = wxLocale::GetSystemEncodingName().Upper();
+
+        // But don't consider ASCII in this case.
+        if ( !encName.empty() )
+        {
+#if wxUSE_FONTMAP
+            wxFontEncoding enc = wxFontMapperBase::GetEncodingFromName(encName);
+            if ( enc == wxFONTENCODING_DEFAULT )
+#else // !wxUSE_FONTMAP
+            if ( encName == wxT("US-ASCII") )
+#endif // wxUSE_FONTMAP/!wxUSE_FONTMAP
+            {
+                // This means US-ASCII when returned from GetEncodingFromName().
+                encName.clear();
+            }
+        }
+#endif // wxUSE_INTL
+
+        // (3) finally use UTF-8 by default
+        if ( encName.empty() )
+            encName = wxT("UTF-8");
+        wxSetEnv(wxT("G_FILENAME_ENCODING"), encName);
+    }
+
+    static wxConvBrokenFileNames fileconv(encName);
+    wxConvFileName = &fileconv;
+#endif // __UNIX__
+}
+#endif // __WXGTK__
+
 int main(int argc, char **argv)
 {
     // tests can be ran non-interactively so make sure we don't show any assert
@@ -419,6 +472,10 @@ int main(int argc, char **argv)
     _CrtSetReportHook(TestCrtReportHook);
 #endif // wxUSE_VC_CRTDBG
 
+#ifdef __WXGTK__
+    SetEncoding();
+#endif // __WXGTK__
+
     try
     {
         return wxEntry(argc, argv);
-- 
1.9.5.msysgit.1

